### PR TITLE
MAINT: Ensure that a copy of the array is returned when calling `block`.

### PR DESCRIPTION
**Pull request by hmaarrfk**
In the case that a user calls
np.block(array),
the old block function would return a view into the array.

This ensures that a new array is returned
@eric-wieser 

~~In thinking about this, I realized that we can probably use this to our advantage and just avoid calling concatenate (which forcibly copies the array).~~

~~This should help the performance when somebody calls `np.block([[[[[a]]]]])`.~~

Developed in light of recent benchmarking results discussed in: https://github.com/numpy/numpy/pull/11971#issuecomment-422508525

### CODE REVIEW

File: numpy/core/tests/test_shape_base.py [hunk 1]
```diff
@@ -191,6 +191,16 @@ def test_2D_array2(self):
 
 
 class TestConcatenate(object):
+    def test_returns_copy(self):
+        a = np.eye(3)
+        b = np.concatenate([a])
+        b[0, 0] = 2
+        assert b[0, 0] != a[0, 0]
+
+        b = np.concatenate(a)
```
Comment (eric-wieser) [inline numpy/core/tests/test_shape_base.py]: This isn't really a good test - unlike `block`, `concatenate` isn't designed to take a single array as an argument.

Comment (hmaarrfk) [reply numpy/core/tests/test_shape_base.py]: ok.

File: numpy/core/shape_base.py [hunk 2]
```diff
@@ -447,8 +447,12 @@ def block_recursion(self, arrays, depth=0):
             # We've 'bottomed out' - arrays is either a scalar or an array
             # type(arrays) is not list
             return atleast_nd(arrays, result_ndim)
-
-    return block_recursion(arrays)
+    if max_depth == 0:
+        # Catch an edge case where the user provies a single array to
+        # block. block should return a copy of that array and not a view.
+        return arrays.copy()
```
Comment (eric-wieser) [inline numpy/core/shape_base.py]: For consistency with the other code path, it probably makes sense to call `block_recursion(arrays)` anyway

Comment (hmaarrfk) [reply numpy/core/shape_base.py]: so `return block_recursion(arrays.copy())`

Comment (eric-wieser) [reply numpy/core/shape_base.py]: `block_recursion(arrays).copy()` would be my choice

Comment (hmaarrfk) [reply numpy/core/shape_base.py]: ok.

**Comment by hmaarrfk** [comment]
@eric-wieser sorry, I reorganized the code in light of your suggesting of forcibly calling `block_recursion`

**Comment by tylerjereddy** [comment]
Nevermind, seems to be in one of my PRs now too -- argh!

**Comment by hmaarrfk** [comment]
@tylerjereddy thanks, I rebased, lets see what happens.

**Comment by tylerjereddy** [comment]
@hmaarrfk It must be something else going on as I've seen it on one of my PRs too..

**Comment by hmaarrfk** [comment]
@tylerjereddy honestly, I hadn't included a test in my original optimization. So codecov forced me to think about it. not so bad.

File: numpy/core/shape_base.py [hunk 3]
```diff
@@ -455,13 +455,31 @@ def atleast_nd(a, ndim):
...
+    # Catch an edge case where the user provies a single array to block.
+    # The user may have passed the array deep inside a nested list.
+    first_array = arrays
+    for _ in range(max_depth):
+        first_array = first_array[0]
+    # The user may have provided an array that is already of the correct shape
+    # Therefore, the `first_array is result`
+    # Otherwise, the `first_array` may have been upcast by `atleast_nd`
+    # In which case, we need to check `result.base`
+    if (first_array is result or first_array is result.base):
```
Comment (eric-wieser) [inline numpy/core/shape_base.py]: Inspecting `.base` like this is rarely a good idea.

**Review by eric-wieser — COMMENTED** [review]
Can we cut back on the scope creep here, and just change the behavior, rather than sneaking in micro-optimizations too?

**Comment by hmaarrfk** [comment]
Of course, let me revert.

File: numpy/core/shape_base.py (hunk 3 above)
Comment (hmaarrfk) [reply numpy/core/shape_base.py]: Is there a good way to check if two arrays point to the same data?

Comment (eric-wieser) [reply numpy/core/shape_base.py]: `np.may_share_memory`, but it's slow. The best way in this case is just to keep track of it yourself.

Comment (hmaarrfk) [reply numpy/core/shape_base.py]: got it.

**Comment by hmaarrfk** [comment]
@mattip If you meant should this PR be merged independently. I think yes. This PR contains some tests specific to the copying behaviour.

**Comment by mattip** [comment]
OK, looks ready to merge. @eric-wieser Any thoughts?

**Comment by mattip** [comment]
Thanks @hmaarrfk
